[SuperEditor][SuperTextField][Android] Show toolbar upon tap on collapsed handle (Resolves #1737)

angelosilvestre opened the pull request:
[SuperEditor][SuperTextField][Android] Show toolbar upon tap on collapsed handle. Resolves #1737

On Android, tapping on the collapsed drag handle doesn't show the popover toolbar.

This PR changes the editor/textfield to show a toolbar containing only the options "Select All" and "Paste" after tapping on the collapsed handle:

[toolbar.webm](https://github.com/superlistapp/super_editor/assets/7597082/b9b66419-3575-41e9-b3eb-5882e620ca73)

Currently, the toolbar builders don't receive the document selection, so I'm using a workaround right now. We need to decide how we want to provide this information to the toolbar builders (for both the editor and the textfield) before merging this PR.

matthew-carroll commented on super_editor/lib/src/default_editor/super_editor.dart:
```diff
@@ -900,8 +900,12 @@ class DefaultAndroidEditorToolbar extends StatelessWidget {
       floatingToolbarKey: floatingToolbarKey,
-      onCopyPressed: _copy,
-      onCutPressed: _cut,
+      onCopyPressed: editorOps.composer.selection == null || !editorOps.composer.selection!.isCollapsed //
```
I'm noticing that the toolbars know how to take editor actions. They use `editorOps`. Can we add a `ValueNotifier<DocumentSelection?>` to the property list for the toolbars to make the decision about which buttons to show?

angelosilvestre replied:
This toolbar is also used for `SuperTextField`, so we can't take anything specific to documents.

matthew-carroll replied:
If that's true, then how did `editorOps` get in here? Those are specific to `SuperEditor`...

angelosilvestre replied:
`DefaultAndroidEditorToolbar` is only used by `SuperEditor`, but `AndroidTextEditingFloatingToolbar` is also used by `SuperTextField`.

matthew-carroll replied:
Ok, help me out here. Without needing to go read all of that code, how are editor operations getting triggered by a toolbar that might be used for both `SuperEditor` and `SuperTextField`.

angelosilvestre replied:
The `AndroidTextEditingFloatingToolbar` receives callbacks for `onCut`, `onCopy`, etc.

matthew-carroll replied:
Does `DefaultAndroidEditorToolbar` use an `AndroidTextEditingFloatingToolbar` internally?

angelosilvestre replied:
Yes

matthew-carroll replied:
Ok....so if `DefaultAndroidEditorToolbar` is specific to `SuperEditor`, then we can provide `DefaultAndroidEditorToolbar` with a `ValueNotifier<Selection>`, right? And then `DefaultAndroidEditorToolbar` can decide whether or not to pass callbacks based on that selection. Am I missing something?

angelosilvestre replied:
Ok, updated.

matthew-carroll commented on super_editor/lib/src/default_editor/super_editor.dart:
```diff
@@ -890,18 +894,24 @@ class DefaultAndroidEditorToolbar extends StatelessWidget {
   final Key? floatingToolbarKey;
   final CommonEditorOperations editorOps;
   final SuperEditorAndroidControlsController editorControlsController;
+  final ValueListenable<DocumentSelection?> selectionNotifier;
```
Do we need to listen to the selection changes or no?

matthew-carroll commented on super_editor/lib/src/infrastructure/touch_controls.dart:
```diff
@@ -16,10 +16,14 @@ enum HandleType {
   /// Represented as global coordinates.
   final Offset focalPoint;
+
+  // TODO: change this.
```
Please do something with this TODO

matthew-carroll commented on super_editor/lib/src/super_textfield/android/android_textfield.dart:
```diff
@@ -658,27 +658,31 @@ Widget _defaultAndroidToolbarBuilder(
-
-      Clipboard.setData(ClipboardData(text: selectedText));
-    },
+    onCutPressed: config.collapsedSelection
```
Can we move the implementation of these behaviors to private methods so that this widget tree isn't so difficult to read?

angelosilvestre replied:
Updated.

matthew-carroll commented on super_editor/test/super_editor/mobile/super_editor_android_overlay_controls_test.dart:
```diff
@@ -63,6 +63,30 @@ void main() {
       await tester.pump(kTapMinTime);
     });
 
+    testWidgetsOnAndroid("shows and hides toolbar upon tap on collapsed handle", (tester) async {
```
Can we also add a test to see what happens when the user taps to show the toolbar, and then taps to place the caret somewhere else?

angelosilvestre replied:
Test added.

matthew-carroll commented on super_editor/lib/src/super_textfield/android/android_textfield.dart:
```diff
@@ -657,44 +657,58 @@ Widget _defaultAndroidToolbarBuilder(
   AndroidEditingOverlayController controller,
   ToolbarConfig config,
 ) {
+  final hasCollapsedSelection = controller.textController.selection != TextRange.empty;
```
Is this calculation correct? Why does a non-empty selection mean that the selection is expanded?

Also, looking at the ternaries down below, I think it would be easier to read the condition of the action is in the first branch and `null` is the second branch.

```
onCutPressed: isSelectionExpanded
  ? () => _onToolbarCutPressed(controller)
  : null,
```

angelosilvestre replied:
Updated.

matthew-carroll commented on super_editor/lib/src/super_textfield/android/android_textfield.dart:
```diff
@@ -657,44 +657,58 @@ Widget _defaultAndroidToolbarBuilder(
-      if (selection.isCollapsed) {
-        return;
-      }
+void _onToolbarCutPressed(AndroidEditingOverlayController controller) {
```
Please move these methods up above the build methods.

matthew-carroll replied:
No, nevermind

matthew-carroll commented on super_editor/test/super_editor/mobile/super_editor_android_overlay_controls_test.dart:
```diff
@@ -63,6 +63,53 @@ void main() {
+      expect(SuperEditorInspector.isMobileToolbarVisible(), isFalse);
+    });
+
+    testWidgetsOnAndroid("hides toolbar when placing the caret", (tester) async {
```
This is more about moving the caret. Let's call it "hides toolbar when the user taps to move the caret"

matthew-carroll commented on super_editor/test/super_textfield/android/super_textfield_android_selection_test.dart:
```diff
@@ -54,6 +55,33 @@ void main() {
       expect(find.byType(AndroidTextEditingFloatingToolbar), findsNothing);
     });
+
+    testWidgetsOnAndroid("tapping at collapsed handle shows/hides the toolbar", (tester) async {
```
Should there also be a test for moving the caret, like the editor tests?

angelosilvestre replied:
Test added

angelosilvestre commented:
@matthew-carroll While I was writing a test for this comment https://github.com/superlistapp/super_editor/pull/1745#discussion_r1445535380 I noticed a weird failure, the toolbar was being displayed after placing a collapsed selection.

After some debugging I noticed a problem in `_onTapUp`. We have this code:

```dart
final previousSelection = widget.textController.selection;
final didTapOnExistingSelection = previousSelection.isCollapsed
    ? tapTextPosition == previousSelection.extent
    : tapTextPosition.offset >= previousSelection.start && tapTextPosition.offset <= previousSelection.end;
```

Although we are trying to inspect the previous selection, at this point of the code we already changed the selection.

Also, looking at the code suggests that tapping on an existing selection should toggle the toolbar, but it doesn't seems that this is happening.

matthew-carroll commented:
@angelosilvestre for your last comment about the bug you found, can you please phrase that as a question for me to answer, or as a description of how you fixed it? Providing only a description of a problem doesn't really give us a path forward.

angelosilvestre commented:
@matthew-carroll The question is, should I fix this issue in this PR? Or should it be a separate ticket?

matthew-carroll commented:
Please fix it here

angelosilvestre commented:
@matthew-carroll Pushed changes to fix the other issue. Here's some changes made:

We have the following code

```dart
void _onTapDown(TapDownDetails details) {
  ...

  // When the user drags, the toolbar should not be visible.
  // A drag can begin with a tap down, so we hide the toolbar
  // preemptively.
  widget.editingOverlayController.hideToolbar();

  ...
}
```

We are always hiding the toolbar on tap down. I removed this code and tested drag selection and the toolbar correctly disappears. 

We also have some code the **should** be toggling the toolbar even when we tap on an expanded selection. However, I noticed that in `SuperEditor`, tapping on an expanded selection places the caret at the tap location instead of toggling the toolbar. I modified to have the same behavior. If this is really what we want, we could simplify the following code to look just for collapsed selections.

```dart
final didTapOnExistingSelection = previousSelection.isCollapsed
    ? tapTextPosition == previousSelection.extent
    : tapTextPosition.offset >= previousSelection.start && tapTextPosition.offset <= previousSelection.end;
```

matthew-carroll commented:
> We are always hiding the toolbar on tap down. I removed this code and tested drag selection and the toolbar correctly disappears.

Did you test long press dragging, too?

> However, I noticed that in SuperEditor, tapping on an expanded selection places the caret at the tap location instead of toggling the toolbar.

Please always check the actual standard platform behavior when you find inconsistent behaviors. What happens on standard Android and iOS in these cases?

angelosilvestre commented:
> Did you test long press dragging, too?

Do you mean drag the caret to move the selection? If that's the case, it seems to be working correctly. What I've found is that long press to select a word doesn't seem to be implemented in `SuperTextField`.

> Please always check the actual standard platform behavior when you find inconsistent behaviors. What happens on standard Android and iOS in these cases?

Tested some apps and the standard behavior is the same implemented in `SuperEditor`. Tapping at an expanded selection places the caret, instead of toggling the toolbar.

matthew-carroll commented:
> Do you mean drag the caret to move the selection?

No, the caret isn't a long-press drag. I'm talking about longpressing to initially select a word, and then dragging from there.

> What I've found is that long press to select a word doesn't seem to be implemented in SuperTextField.

Can you please double check that? Do we have open tickets for that? Did we only implement it for `SuperEditor`? Text fields should definitely support it.

angelosilvestre commented:
> Can you please double check that? Do we have open tickets for that? Did we only implement it for SuperEditor? Text fields should definitely support it.

I double checked and it isn't implemented. The original PR only adds this gesture to `SuperEditor` and `SuperReader`: https://github.com/superlistapp/super_editor/pull/1489

I didn't find any open tickets for the textfield long-press.

matthew-carroll commented on super_editor/lib/src/super_textfield/android/_user_interaction.dart:
```diff
@@ -160,24 +160,11 @@ class AndroidTextFieldTouchInteractorState extends State<AndroidTextFieldTouchIn
-    // preemptively.
-    widget.editingOverlayController.hideToolbar();
-  }
+  void _onTapDown(TapDownDetails details) {}
```
Why do we have an empty function?

angelosilvestre replied:
I forgot to removed. Updated.

matthew-carroll reviewed (approved):
LGTM
